dali/operators/imgcodec/decoder.cc

```diff
@@ -0,0 +1,411 @@
+interpreted as absolute or relative coordinates, depending on the value of
+``normalized_shape``.)code");
+
+
+DALI_SCHEMA(experimental__decoders__ImageRandomCrop)
+  .DocStr(R"code(Decodes images and randomly crops them.
+
+The cropping window's area (relative to the entire image) and aspect ratio can be restricted to
+a range of values specified by ``area`` and ``aspect_ratio`` arguments, respectively.
+
+When possible, the operator uses the ROI decoding APIs (for example, *libjpeg-turbo* and *nvJPEG*)
+to reduce the decoding time and memory usage. When the ROI decoding is not supported for a given
+image format, it will decode the entire image and crop the selected ROI.
+
+The output of the decoder is in *HWC* layout.
+
+Supported formats: JPG, BMP, PNG, TIFF, PNM, PPM, PGM, PBM, JPEG 2000, WebP.
+
+.. note::
+  JPEG 2000 region-of-interest (ROI) decoding is not accelerated on the GPU, and will use
+  a CPU implementation regardless of the selected backend. For a GPU accelerated implementation,
+  consider using separate ``decoders.image`` and ``random_crop`` operators.
+
+.. note::
+  EXIF orientation metadata is used to rectify the image.)code")
+  .NumInput(1)
+  .NumOutput(1)
+  .AddParent("ImgcodecDecoderAttr")
+  .AddParent("RandomCropAttr");
+
+
+bool ImgcodecHostDecoder::SetupImpl(std::vector<OutputDesc> &output_descs,
+                            const workspace_t<CPUBackend> &ws) {
+  SetupShapes(spec_, ws, output_descs, ws.GetThreadPool());
+  return true;
+}
+
+void ImgcodecHostDecoder::RunImpl(workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.Input<CPUBackend>(0);
+  auto &output = ws.Output<CPUBackend>(0);
+  output.SetLayout("HWC");
+
+  auto *decoder = GetDecoderInstance();
+
+  int nsamples = input.shape().num_samples();
+
+  DecodeContext ctx;
+  ctx.tp = &ws.GetThreadPool();
+
+  BatchVector<ImageSource> srcs;
+  BatchVector<ImageSource *> src_ptrs;
+  srcs.resize(nsamples);
+  src_ptrs.resize(nsamples);
+
+  for (int i = 0; i < nsamples; i++) {
+    srcs[i] = SampleAsImageSource(input[i]);
+    src_ptrs[i] = &srcs[i];
+  }
+
+  auto results = decoder->Decode(ctx, output, make_span(src_ptrs), opts_, make_span(rois_));
+  for (const auto &result : results) {
+    if (!result.success) {
+      std::rethrow_exception(result.exception);
+    }
+  }
+}
+
+int ImgcodecHostDecoder::GetDeviceId() {
+  return CPU_ONLY_DEVICE_ID;
+}
+
+
+bool ImgcodecMixedDecoder::SetupImpl(std::vector<OutputDesc> &output_descs,
+                                     const workspace_t<MixedBackend> &ws) {
+  SetupShapes(spec_, ws, output_descs, thread_pool_);
+  return true;
+}
+
+void ImgcodecMixedDecoder::Run(workspace_t<MixedBackend> &ws) {
+  const auto &input = ws.Input<CPUBackend>(0);
+  auto &output = ws.Output<GPUBackend>(0);
+  output.SetLayout("HWC");
+
+  auto *decoder = GetDecoderInstance();
+
+  int nsamples = input.shape().num_samples();
+
+  DecodeContext ctx;
+  ThreadPool tp(4, GetDeviceId(), false, "Mixed decoder");
```

Comment:
Shouldn't that be more persistent? Shouldn't that be just the `ImgcodecMixedDecoder::thread_pool_`? Or is it intentional choice to work around some limitations of the existing thread pool that were mentioned lately?